handler/src/main/java/io/netty/handler/ssl/AbstractSniHandler.java

```diff
@@ -0,0 +1,323 @@
+
+                                            try {
+                                                select(ctx, IDN.toASCII(hostname,
+                                                                        IDN.ALLOW_UNASSIGNED).toLowerCase(Locale.US));
+                                            } catch (Throwable t) {
+                                                PlatformDependent.throwException(t);
+                                            }
+                                            return;
+                                        } else {
+                                            // invalid enum value
+                                            break loop;
+                                        }
+                                    }
+
+                                    offset += extensionLength;
+                                }
+                            }
+                            // Fall-through
+                        default:
+                            //not tls, ssl or application data, do not try sni
+                            break loop;
+                    }
+                }
+            } catch (Throwable e) {
+                // unexpected encoding, ignore sni and use default
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Unexpected client hello packet: " + ByteBufUtil.hexDump(in), e);
+                }
+            }
+            // Just select the default SslContext
+            select(ctx, null);
+        }
+    }
+
+    private void select(final ChannelHandlerContext ctx, final String hostname) throws Exception {
+        Future<T> future = lookup(ctx, hostname);
+        if (future.isDone()) {
+            onLookupComplete(ctx, hostname, future);
+        } else {
+            suppressRead = true;
+            future.addListener(new FutureListener<T>() {
+                @Override
+                public void operationComplete(Future<T> future) throws Exception {
+                    try {
+                        suppressRead = false;
+                        try {
+                            onLookupComplete(ctx, hostname, future);
+                        } catch (DecoderException err) {
+                            ctx.fireExceptionCaught(err);
+                        } catch (Throwable cause) {
+                            ctx.fireExceptionCaught(new DecoderException(cause));
+                        }
+                    } finally {
+                        if (readPending) {
+                            readPending = false;
+                            ctx.read();
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Kicks off a lookup for the given SNI value and returns a {@link Future} which in turn will
+     * notify the {@link #onLookupValue(ChannelHandlerContext, String, Future)} on completion.
+     *
+     * @see #onLookupValue(ChannelHandlerContext, String, Future)
```

Comment:
This method reference cannot be found. Is `#onLookupValue(ChannelHandlerContext, String, Object)` or `#onLookupComplete(ChannelHandlerContext, String, Future)` what you wanted? Also fix the reference above.